metals/src/main/scala/scala/meta/internal/worksheets/MetalsWorksheetProvider.scala

```diff
@@ -0,0 +1,352 @@
+              thread.stop()
+            }
+          })
+        )
+        stopThreadOnCancel(path, result, thread)
+        thread.start()
+        thread.join()
+      }
+      jobs.submit(
+        result,
+        () => {
+          try runEvaluation()
+          catch {
+            case NonFatal(_) | InterruptException() =>
+              () // allow job queue to process next worksheet evaluation request.
+          }
+        }
+      )
+      result.asScala.recover {
+        case NonFatal(_) | InterruptException() =>
+          // Clear all decorations when evaluation fails.
+          Array.empty
+      }
+    }
+  }
+
+  /**
+   * Prompts the user to cancel the task after a few seconds.
+   *
+   * Attempts to gracefully shut down the thread when users requests to cancel:
+   * First tries `Thread.interrupt()` with fallback to `Thread.stop()` after
+   * one second if interruption doesn't work.
+   */
+  private def stopThreadOnCancel(
+      path: AbsolutePath,
+      result: CompletableFuture[Array[DecorationOptions]],
+      thread: Thread
+  ): Unit = {
+    // Last resort, if everything else fails we use `Thread.stop()`.
+    val stopThread = new Runnable {
+      def run(): Unit = {
+        if (thread.isAlive()) {
+          scribe.warn(s"thread stop: ${thread.getName()}")
+          Cancelable.cancelAll(
+            List(
+              Cancelable(() => thread.stop()),
+              cancelables
+            )
+          )
+        }
+      }
+    }
+    // If the program is running for more than
+    // `userConfig().worksheetCancelTimeout`, then display a prompt for the user
+    // to cancel the program.
+    val promptUserToCancel = new Runnable {
+      def run(): Unit = {
+        if (!result.isDone()) {
+          val cancel = languageClient.metalsSlowTask(
+            new MetalsSlowTaskParams(
+              s"Evaluating worksheet '${path.filename}'",
+              noLogs = true
+            )
+          )
+          cancel.asScala.foreach { c =>
+            if (c.cancel && thread.isAlive()) {
+              // User has requested to cancel a running program. first line of
+              // defense is `Thread.interrupt()`. Fingers crossed it's enough.
+              result.complete(Array.empty)
+              sh.schedule(stopThread, 1, TimeUnit.SECONDS)
+              scribe.warn(s"thread interrupt: ${thread.getName()}")
+              thread.interrupt()
+            }
+          }
+          result.asScala.onComplete(_ => cancel.cancel(true))
+        }
+      }
+    }
+    sh.schedule(
+      promptUserToCancel,
+      userConfig().worksheetCancelTimeout,
+      TimeUnit.SECONDS
+    )
+  }
+
+  private def evaluateWorksheet(
+      path: AbsolutePath,
+      token: CancelToken
+  ): Array[DecorationOptions] = {
+    val commentHeader = " // "
+    val input = path.toInputFromBuffers(buffers)
+    val decorations = for {
+      ctx <- getContext(path)
+      source <- dialects.Sbt1(input).parse[Source].toOption
```

Comment:
I understand the logic but seeing the `Sbt1` dialect here is a bit confusing. Can we have a `Worksheet` dialect instead? It can be a type alias, it's purely for readability.

Reply:
I added a `Worksheet` alias. We could consider adding this to Scalameta itself.